presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/TransformFilteringSemiJoinToInnerJoin.java

```diff
@@ -0,0 +1,168 @@
+ * - Filter (semiJoinSymbol AND predicate)
+ *    - SemiJoin (semiJoinSymbol <- (a IN b))
+ *        source: plan A producing symbol a
+ *        filtering source: plan B producing symbol b
+ * </pre>
+ * <p/>
+ * Into:
+ * <pre>
+ * - Filter (predicate)
+ *    - Project (semiJoinSymbol <- TRUE)
+ *       - Join INNER on (a = b)
+ *          - source
+ *          - Aggregation distinct(b)
+ *             - filtering source
+ * </pre>
+ */
+public class TransformFilteringSemiJoinToInnerJoin
+        implements Rule<FilterNode>
+{
+    private static final Capture<SemiJoinNode> SEMI_JOIN = newCapture();
+
+    private static final Pattern<FilterNode> PATTERN = filter()
+            .with(source().matching(semiJoin().capturedAs(SEMI_JOIN)));
+
+    @Override
+    public Pattern<FilterNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isRewriteFilteringSemiJoinToInnerJoin(session);
+    }
+
+    @Override
+    public Result apply(FilterNode filterNode, Captures captures, Context context)
+    {
+        SemiJoinNode semiJoin = captures.get(SEMI_JOIN);
+
+        // Do not transform semi-join in context of DELETE
+        if (PlanNodeSearcher.searchFrom(semiJoin.getSource(), context.getLookup())
+                .where(node -> node instanceof TableScanNode && ((TableScanNode) node).isForDelete())
+                .matches()) {
+            return Result.empty();
+        }
+
+        Symbol semiJoinSymbol = semiJoin.getSemiJoinOutput();
+        Predicate<Expression> isSemiJoinSymbol = expression -> expression.equals(semiJoinSymbol.toSymbolReference());
+
+        List<Expression> conjuncts = extractConjuncts(filterNode.getPredicate());
+        if (conjuncts.stream().noneMatch(isSemiJoinSymbol)) {
+            return Result.empty();
+        }
+        Expression filteredPredicate = and(conjuncts.stream()
+                .filter(not(isSemiJoinSymbol))
+                .collect(toImmutableList()));
+
+        PlanNode filteringSourceDistinct = new AggregationNode(
+                context.getIdAllocator().getNextId(),
+                semiJoin.getFilteringSource(),
+                ImmutableMap.of(),
+                singleGroupingSet(ImmutableList.of(semiJoin.getFilteringSourceJoinSymbol())),
+                ImmutableList.of(),
+                SINGLE,
+                Optional.empty(),
+                Optional.empty());
+
+        Expression joinFilter = new ComparisonExpression(EQUAL, semiJoin.getSourceJoinSymbol().toSymbolReference(), semiJoin.getFilteringSourceJoinSymbol().toSymbolReference());
+
+        JoinNode innerJoin = new JoinNode(
+                semiJoin.getId(),
+                INNER,
+                semiJoin.getSource(),
+                filteringSourceDistinct,
+                ImmutableList.of(),
+                semiJoin.getSource().getOutputSymbols(),
+                ImmutableList.of(),
+                Optional.of(joinFilter),
```

Comment:
why not to put `filteredPredicate` immediatelly?

Reply:
`filteredPredicate` might depend on the semi join symbol.